vm/devices/get/guest_emulation_device/src/lib.rs

```diff
@@ -613,33 +615,64 @@ impl<T: RingMem + Unpin> GedChannel<T> {
         Ok(())
     }
 
-    fn handle_vmgs_get_device_info(&mut self) -> Result<(), Error> {
-        let response = get_protocol::VmgsGetDeviceInfoResponse::new(
-            VmgsIoStatus::SUCCESS,
-            VMGS_CAPACITY as u64,
-            VMGS_SECTOR_SIZE as u16,
-            VMGS_SECTOR_SIZE as u16,
-            512,
-        );
+    fn handle_vmgs_get_device_info(
+        &mut self,
+        state: &mut GuestEmulationDevice,
+    ) -> Result<(), Error> {
+        let response = if let Some(vmgs) = &state.vmgs {
+            get_protocol::VmgsGetDeviceInfoResponse::new(
+                VmgsIoStatus::SUCCESS,
+                vmgs.disk.sector_count(),
+                vmgs.disk.sector_size() as u16,
+                vmgs.disk.physical_sector_size() as u16,
+                vmgs.buf.len() as u32,
+            )
+        } else {
+            get_protocol::VmgsGetDeviceInfoResponse::new(VmgsIoStatus::DEVICE_ERROR, 0, 0, 0, 0)
+        };
         self.channel
             .try_send(response.as_bytes())
             .map_err(Error::Vmbus)?;
         Ok(())
     }
 
-    fn handle_vmgs_read(
+    async fn handle_vmgs_read(
         &mut self,
         state: &mut GuestEmulationDevice,
         message_buf: &[u8],
     ) -> Result<(), Error> {
         let message = get_protocol::VmgsReadRequest::read_from_prefix(message_buf)
             .ok_or(Error::MessageTooSmall)?;
 
-        let (status, payload) = if let Some(vmgs) = state.vmgs(message.offset, message.length) {
-            if vmgs.len() > MAX_PAYLOAD_SIZE {
+        let (status, payload) = if let Some(vmgs) = &mut state.vmgs {
+            let len = message.length as u64 * vmgs.disk.sector_size() as u64;
+            if len > MAX_PAYLOAD_SIZE as u64 {
                 return Err(Error::InvalidFieldValue);
             }
-            (VmgsIoStatus::SUCCESS, &*vmgs)
+
+            // FUTURE: since this IO may take a long time, consider storing the
```

Comment:
Just double checking - the suggestion isn't to make vmgs read/write over the GET actually async, right? That would be a departure from the existing GET's (unfortunate, pre-existing) blocking semantics.

Reply:
Right, the suggestion is just to make sure we don't block host activity (pausing VMs, inspecting the GET, etc.).

Reply:
might be worth tweaking some of the phrasing to make that intent more clear